### PR TITLE
feat: monitor deployed versions of our sandbox operators

### PR DESCRIPTION
**Pull request by mfrancisc**
Add check for deployed version of member operator. The check is enabled only for production environments where a preconfigured secret was deployed by a sys admin.

If the deployed commit and the latest GitHub commit are not matching the memberstatus will be set to unhealthy.

Jira: https://issues.redhat.com/browse/SANDBOX-53

Related PRs:
- api: https://github.com/codeready-toolchain/api/pull/366
- host-operator follow up: https://github.com/codeready-toolchain/host-operator/pull/832 
- toolchain-common: https://github.com/codeready-toolchain/toolchain-common/pull/315
- host-operator: https://github.com/codeready-toolchain/host-operator/pull/819

Paired with:
- toolchain-e2e: https://github.com/codeready-toolchain/toolchain-e2e/pull/745

### CODE REVIEW

**Review by MatousJobanek — COMMENTED** [review]
I have the same concerns as mentioned in the host-operator PR https://github.com/codeready-toolchain/host-operator/pull/819#pullrequestreview-1457520605

File: controllers/memberstatus/memberstatus_controller_test.go [hunk 1]
```diff
@@ -61,7 +67,7 @@ func TestNoMemberStatusFound(t *testing.T) {
 		// given
 		requestName := "bad-name"
 		getHostClusterFunc := newGetHostClusterReady
-		reconciler, req, _ := prepareReconcile(t, requestName, getHostClusterFunc, allNamespacesCl)
+		reconciler, req, _ := prepareReconcile(t, requestName, getHostClusterFunc, allNamespacesCl, mockLastGitHubAPICall, test.MockGitHubClientForRepositoryCommits(buildCommitSHA, time.Now().Add(-time.Hour*1)))
```
Comment (MatousJobanek) [inline controllers/memberstatus/memberstatus_controller_test.go:70]: You could do the same as in host-operator PR https://github.com/codeready-toolchain/host-operator/pull/819#discussion_r1243552199

File: controllers/memberstatus/memberstatus_controller_test.go [hunk 2]
```diff
@@ -113,7 +119,7 @@ func TestOverallStatusCondition(t *testing.T) {
 		memberOperatorDeployment := newMemberDeploymentWithConditions(status.DeploymentAvailableCondition(), status.DeploymentProgressingCondition())
 		memberStatus := newMemberStatus()
 		getHostClusterFunc := newGetHostClusterReady
-		reconciler, req, fakeClient := prepareReconcile(t, requestName, getHostClusterFunc, allNamespacesCl, append(nodeAndMetrics, memberOperatorDeployment, memberStatus)...)
+		reconciler, req, fakeClient := prepareReconcile(t, requestName, getHostClusterFunc, allNamespacesCl, mockLastGitHubAPICall, test.MockGitHubClientForRepositoryCommits(buildCommitSHA, time.Now().Add(-time.Hour*1)), append(nodeAndMetrics, memberOperatorDeployment, memberStatus)...)
 
 		// when
 		res, err := reconciler.Reconcile(context.TODO(), req)
```
Comment (MatousJobanek) [inline controllers/memberstatus/memberstatus_controller_test.go:125]: for these happy paths `t.Run("All components ready", func(t *testing.T) {` shouldn't we update also the assertion so it expects the resource revision to be set? Just asking :wink:

File: controllers/memberstatus/memberstatus_controller_test.go [hunk 3]
```diff
@@ -668,6 +674,122 @@ func TestOverallStatusCondition(t *testing.T) {
...
+		// given
+		config := commonconfig.NewMemberOperatorConfigWithReset(t, testconfig.MemberEnvironment("prod"), testconfig.MemberStatus().GitHubSecretRef("github").GitHubSecretAccessTokenKey("accessToken"))
+		requestName := defaultMemberStatusName
+		memberOperatorDeployment := newMemberDeploymentWithConditions(status.DeploymentAvailableCondition(), status.DeploymentProgressingCondition())
+		memberStatus := newMemberStatus()
+		getHostClusterFunc := newGetHostClusterReady
+		// we have a secret that contains the access token for GitHub authenticated APIs
+		githubSecret := test.CreateSecret("github", test.MemberOperatorNs, map[string][]byte{
+			"accessToken": []byte("abcd1234"),
+		})
```
Comment (MatousJobanek) [inline controllers/memberstatus/memberstatus_controller_test.go:718]: the given parts in the test seems to be the same - could we move it to a parent test so it could be reused?

File: controllers/memberstatus/memberstatus_controller_test.go [hunk 4]
```diff
@@ -668,6 +674,122 @@ func TestOverallStatusCondition(t *testing.T) {
...
+			res, err := reconciler.Reconcile(context.TODO(), req)
+
+			// then
+			require.NoError(t, err)
+			assert.Equal(t, requeueResult, res)
+			AssertThatMemberStatus(t, req.Namespace, requestName, fakeClient).
+				HasCondition(ComponentsReady()).                                                                     // all components are ready
+				HasMemberOperatorConditions(ConditionReady(toolchainv1alpha1.ToolchainStatusDeploymentReadyReason)). // we have only one condition for the deployment status, no condition for the version match
+				HasMemoryUsage(OfNodeRole("master", 33), OfNodeRole("worker", 25)).
+				HasRoutes("https://console.member-cluster/console/", "https://codeready-codeready-workspaces-operator.member-cluster/che/", routesAvailable())
```
Comment (MatousJobanek) [inline controllers/memberstatus/memberstatus_controller_test.go:763]: I can be blind, but I don't see the assertion of the revision check condition something like that:
```go
HasMemberOperatorRevisionCheckConditions(ConditionReady(toolchainv1alpha1.ToolchainStatusDeploymentRevisionCheckDisabledReason ...
```

File: controllers/memberstatus/memberstatus_controller_test.go [hunk 5]
```diff
@@ -668,6 +674,122 @@ func TestOverallStatusCondition(t *testing.T) {
...
```
Comment (MatousJobanek) [inline controllers/memberstatus/memberstatus_controller_test.go:789]: the same - I'm missing the verification that the condition is set

**Review by MatousJobanek — COMMENTED** [review]
the controller code looks OK. Just a few comments to the unit tests

File: controllers/memberstatus/memberstatus_controller_test.go (hunk 1 above)
Comment (mfrancisc) [reply controllers/memberstatus/memberstatus_controller_test.go]: I've cleaned it up a bit . Pls see [78a30e1](https://github.com/codeready-toolchain/member-operator/pull/446/commits/78a30e10315b864af392e7a11c435de06468920a) and [daee1ad](https://github.com/codeready-toolchain/member-operator/pull/446/commits/daee1ad8962b8530c1da5e096e0f3ac12a8b7a1d)

File: controllers/memberstatus/memberstatus_controller_test.go (hunk 2 above)
Comment (mfrancisc) [reply controllers/memberstatus/memberstatus_controller_test.go]: You are right, I've missed those, I've now added the expected disabled condition. I was thinking that maybe at some point I can remove the specific test that verifies if the component is up to date _"member operator deployment version is up to date"_ and have this check in _"All components ready"_. For now I did it the same way as host-operator, but probably is worth refactoring it.

Plz see [78a30e1](https://github.com/codeready-toolchain/member-operator/pull/446/commits/78a30e10315b864af392e7a11c435de06468920a) and [daee1ad](https://github.com/codeready-toolchain/member-operator/pull/446/commits/daee1ad8962b8530c1da5e096e0f3ac12a8b7a1d)

File: controllers/memberstatus/memberstatus_controller_test.go (hunk 3 above)
Comment (mfrancisc) [reply controllers/memberstatus/memberstatus_controller_test.go]: I've created _"member operator deployment revision check"_ as a parent for all the revision check unit tests.
TBH I tried putting also the `devConfig` and `prodConfig` there but the tests were randomly failing , I wasn't yet able to understand why, the reconciler and fakeClient are scoped to each specific test, so not sure what's happening. I've tried also moving the `memberStatus` to each test but still not working. 

Long story short - I had to create this `devConfig` and `prodConfig` in each subtest.

Plz see [78a30e1](https://github.com/codeready-toolchain/member-operator/pull/446/commits/78a30e10315b864af392e7a11c435de06468920a) and [daee1ad](https://github.com/codeready-toolchain/member-operator/pull/446/commits/daee1ad8962b8530c1da5e096e0f3ac12a8b7a1d)

File: controllers/memberstatus/memberstatus_controller_test.go (hunk 4 above)
Comment (mfrancisc) [reply controllers/memberstatus/memberstatus_controller_test.go]: I are not blind, I forgot to add it back. Since we went from 2 conditions to 1 conditions and now back to 2 conditions , I forgot to update the tests 🤦‍♂️ . Good catch again!

Done in [78a30e1](https://github.com/codeready-toolchain/member-operator/pull/446/commits/78a30e10315b864af392e7a11c435de06468920a) and [daee1ad](https://github.com/codeready-toolchain/member-operator/pull/446/commits/daee1ad8962b8530c1da5e096e0f3ac12a8b7a1d)

File: controllers/memberstatus/memberstatus_controller_test.go (hunk 5 above)
Comment (mfrancisc) [reply controllers/memberstatus/memberstatus_controller_test.go]: Done in [78a30e1](https://github.com/codeready-toolchain/member-operator/pull/446/commits/78a30e10315b864af392e7a11c435de06468920a) and [daee1ad](https://github.com/codeready-toolchain/member-operator/pull/446/commits/daee1ad8962b8530c1da5e096e0f3ac12a8b7a1d)

Thanks!

**Comment by mfrancisc** [comment]
@MatousJobanek thanks for your review. I've did a couple of additional fixes:
- change `HEAD` to master
- initialize `operatorStatus` with revisionCheck field


but while testing locally I'm noticing that I get an unexpected error:

```
"unable to find ConditionReady type in existing conditions. Waiting for next attempt ..."
``` 

I think there is something missing to this new `revisionCheck` section when it's initialized.

**Review by MatousJobanek — APPROVED** [review]
Looks great :+1: :rocket:

File: controllers/memberstatus/memberstatus_controller_test.go [hunk 6]
```diff
@@ -668,6 +679,103 @@ func TestOverallStatusCondition(t *testing.T) {
...
+					HasCondition(ComponentsReady()).                                                                     // all components are ready
+					HasMemberOperatorConditions(ConditionReady(toolchainv1alpha1.ToolchainStatusDeploymentReadyReason)). // we have only one condition for the deployment status
+					HasMemberOperatorRevisionCheckConditions(ConditionReadyWithMessage(toolchainv1alpha1.ToolchainStatusDeploymentRevisionCheckDisabledReason, "is not running in prod environment")).
+					HasMemoryUsage(OfNodeRole("master", 33), OfNodeRole("worker", 25)).
+					HasRoutes("https://console.member-cluster/console/", "https://codeready-codeready-workspaces-operator.member-cluster/che/", routesAvailable())
+			})
+			//
+			t.Run("when environment is prod but github secret is not present", func(t *testing.T) {
+				// given
+				prodConfig := commonconfig.NewMemberOperatorConfigWithReset(t, testconfig.MemberEnvironment("prod"), testconfig.MemberStatus().GitHubSecretRef("github").GitHubSecretAccessTokenKey("accessToken"))
```
[the remark this answers is not in the document]
Comment (mfrancisc) [reply controllers/memberstatus/memberstatus_controller_test.go]: The intention was to show that even if the github secret is configured in the config but not present in the cluster the code can still handle it. But if this is confusing I'm removing it, pls see: https://github.com/codeready-toolchain/member-operator/pull/446/commits/628ab4d66d20d24437ff1b27bf64e410db73d3a7

File: controllers/memberstatus/memberstatus_controller_test.go [hunk 7]
```diff
@@ -124,13 +131,14 @@ func TestOverallStatusCondition(t *testing.T) {
 		AssertThatMemberStatus(t, req.Namespace, requestName, fakeClient).
 			HasCondition(ComponentsReady()).
 			HasMemoryUsage(OfNodeRole("master", 33), OfNodeRole("worker", 25)).
+			HasMemberOperatorRevisionCheckConditions(ConditionReadyWithMessage(toolchainv1alpha1.ToolchainStatusDeploymentRevisionCheckDisabledReason, "access token key is not provided")).
```
[the remark this answers is not in the document]
Comment (mfrancisc) [reply controllers/memberstatus/memberstatus_controller_test.go:134]: As mentioned here https://github.com/codeready-toolchain/member-operator/pull/446#discussion_r1247644910 , I was considering in adding the happy path to the "All components ready" test, and remove the specific test "member operator deployment version is up to date". I'll open a PR and refactor this.

**Comment by mfrancisc** [comment]
> Looks great 👍 🚀

Thanks for your review, I've also fixed an issue with the `RevisionCheck` field not being copied between member status updates, please see:  https://github.com/codeready-toolchain/member-operator/pull/446/commits/628ab4d66d20d24437ff1b27bf64e410db73d3a7

I'm afraid host-operator might have a similar issue, I'll open a PR to fix it there as well.
